pdns/xdp-filter-tc.ebpf.src

```diff
@@ -0,0 +1,380 @@
+#include <net/sock.h>
+#include <uapi/linux/udp.h>
+#include <uapi/linux/ip.h>
+#include <uapi/linux/ipv6.h>
+
+#define DNS_PORT      53
+
+// do not use libc includes because this causes clang
+// to include 32bit headers on 64bit ( only ) systems.
+typedef __u8  uint8_t;
+typedef __u16 uint16_t;
+typedef __u32 uint32_t;
+typedef __u64 uint64_t;
+#define memcpy __builtin_memcpy
+
+/*
+ *  Helper pointer to parse the incoming packets
+ */
+struct cursor {
+  void *pos;
+  void *end;
+};
+
+/*
+ *  Store the VLAN header
+ */
+struct vlanhdr {
+  uint16_t tci;
+  uint16_t encap_proto;
+};
+
+/*
+ *  Store the DNS header
+ */
+struct dnshdr {
+  uint16_t id;
+  union {
+  	struct {
+#if BYTE_ORDER == LITTLE_ENDIAN
+  		uint8_t  rd     : 1;
+  		uint8_t  tc     : 1;
+  		uint8_t  aa     : 1;
+  		uint8_t  opcode : 4;
+  		uint8_t  qr     : 1;
+
+  		uint8_t  rcode  : 4;
+  		uint8_t  cd     : 1;
+  		uint8_t  ad     : 1;
+  		uint8_t  z      : 1;
+  		uint8_t  ra     : 1;
+#elif BYTE_ORDER == BIG_ENDIAN || BYTE_ORDER == PDP_ENDIAN
+  		uint8_t  qr     : 1;
+  		uint8_t  opcode : 4;
+  		uint8_t  aa     : 1;
+  		uint8_t  tc     : 1;
+  		uint8_t  rd     : 1;
+
+  		uint8_t  ra     : 1;
+  		uint8_t  z      : 1;
+  		uint8_t  ad     : 1;
+  		uint8_t  cd     : 1;
+  		uint8_t  rcode  : 4;
+#endif
+  	}        as_bits_and_pieces;
+  	uint16_t as_value;
+  } flags;
+  uint16_t qdcount;
+  uint16_t ancount;
+  uint16_t nscount;
+  uint16_t arcount;
+};
+
+/*
+ * Store the qname and qtype
+ */
+struct dns_qname
+{
+  uint8_t qname[255];
+  uint16_t qtype;
+};
+
+BPF_TABLE("hash", uint32_t, uint64_t, v4filter, 1024);
+BPF_TABLE("hash", struct in6_addr, uint64_t, v6filter, 1024);
+BPF_TABLE("hash", struct dns_qname, uint64_t, qnamefilter, 1024);
```

Comment:
Since we are already using a different table format than the existing one, would it make sense to update the value of the map to a struct holding a counter and an action, allowing us to merge the two files, for example something like:
```C
enum dns_action { drop = 0, tc = 1 };
struct map_value
{
  uint64_t counter;
  uint8_t action;
};
```

Reply:
It is a really good idea